xarray/core/indexing.py

```diff
@@ -325,6 +325,18 @@ def as_integer_slice(value):
     return slice(start, stop, step)
 
 
+class IndexCallable:
+    """Provide getitem syntax for a callable object."""
+
+    __slots__ = ("func",)
+
+    def __init__(self, func):
+        self.func = func
+
+    def __getitem__(self, key):
+        return self.func(key)
+
+
```

Comment:
```suggestion
        def __setitem__(self, key):
            raise NotImplementedError
```

This will be needed eventually.